### ISSUE TITLE
Linking test code fails on macOS

### ISSUE THREAD

**Issue by marshallpierce** [briansmith/ring#383]
On my macOS 10.12.2 box, I get the following output from `cargo test --verbose`:

```
   Compiling untrusted v0.3.2
   Compiling lazy_static v0.2.2
   Compiling ring v0.6.0-alpha1 (file:///Users/mbp/dev/forks/ring)
     Running `rustc /Users/mbp/.cargo/registry/src/github.com-1ecc6299db9ec823/untrusted-0.3.2/src/untrusted.rs --crate-name untrusted --crate-type lib -g -C metadata=6d5be7309ea48309 -C extra-filename=-6d5be7309ea48309 --out-dir /Users/mbp/dev/forks/ring/target/debug/deps --emit=dep-info,link -L dependency=/Users/mbp/dev/forks/ring/target/debug/deps --cap-lints allow`
     Running `rustc /Users/mbp/.cargo/registry/src/github.com-1ecc6299db9ec823/lazy_static-0.2.2/src/lib.rs --crate-name lazy_static --crate-type lib -g -C metadata=7f1b96a3a3eb529d -C extra-filename=-7f1b96a3a3eb529d --out-dir /Users/mbp/dev/forks/ring/target/debug/deps --emit=dep-info,link -L dependency=/Users/mbp/dev/forks/ring/target/debug/deps --cap-lints allow`
     Running `rustc build.rs --crate-name build_script_build --crate-type bin -g --cfg feature=\"use_heap\" --cfg feature=\"default\" --cfg feature=\"dev_urandom_fallback\" -C metadata=5ae0aa2ea1993729 --out-dir /Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729 --emit=dep-info,link -L dependency=/Users/mbp/dev/forks/ring/target/debug/deps`
     Running `/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/build-script-build`
     Running `rustc src/lib.rs --crate-name ring --crate-type lib -g --cfg feature=\"use_heap\" --cfg feature=\"default\" --cfg feature=\"dev_urandom_fallback\" -C metadata=5ae0aa2ea1993729 --out-dir /Users/mbp/dev/forks/ring/target/debug/deps --emit=dep-info,link -L dependency=/Users/mbp/dev/forks/ring/target/debug/deps --extern lazy_static=/Users/mbp/dev/forks/ring/target/debug/deps/liblazy_static-7f1b96a3a3eb529d.rlib --extern untrusted=/Users/mbp/dev/forks/ring/target/debug/deps/libuntrusted-6d5be7309ea48309.rlib -L native=/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/out/lib -l static=ring-core -l static=ring-test -l dylib=stdc++`
     Running `rustc src/lib.rs --crate-name ring -g --test --cfg feature=\"use_heap\" --cfg feature=\"default\" --cfg feature=\"dev_urandom_fallback\" -C metadata=2a96e7777e87260e -C extra-filename=-2a96e7777e87260e --out-dir /Users/mbp/dev/forks/ring/target/debug/deps --emit=dep-info,link -L dependency=/Users/mbp/dev/forks/ring/target/debug/deps --extern lazy_static=/Users/mbp/dev/forks/ring/target/debug/deps/liblazy_static-7f1b96a3a3eb529d.rlib --extern untrusted=/Users/mbp/dev/forks/ring/target/debug/deps/libuntrusted-6d5be7309ea48309.rlib -L native=/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/out/lib -l static=ring-core -l static=ring-test -l dylib=stdc++`
     Running `rustc examples/checkdigest.rs --crate-name checkdigest --crate-type bin -g --cfg feature=\"use_heap\" --cfg feature=\"default\" --cfg feature=\"dev_urandom_fallback\" -C metadata=5ae0aa2ea1993729 --out-dir /Users/mbp/dev/forks/ring/target/debug/examples --emit=dep-info,link -L dependency=/Users/mbp/dev/forks/ring/target/debug/deps --extern lazy_static=/Users/mbp/dev/forks/ring/target/debug/deps/liblazy_static-7f1b96a3a3eb529d.rlib --extern untrusted=/Users/mbp/dev/forks/ring/target/debug/deps/libuntrusted-6d5be7309ea48309.rlib --extern ring=/Users/mbp/dev/forks/ring/target/debug/deps/libring.rlib -L native=/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/out/lib`
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-m64" "-L" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "/Users/mbp/dev/forks/ring/target/debug/deps/ring-2a96e7777e87260e.0.o" "-o" "/Users/mbp/dev/forks/ring/target/debug/deps/ring-2a96e7777e87260e" "-Wl,-dead_strip" "-nodefaultlibs" "-L" "/Users/mbp/dev/forks/ring/target/debug/deps" "-L" "/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/out/lib" "-L" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib" "-Wl,-force_load,/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/out/lib/libring-core.a" "-Wl,-force_load,/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/out/lib/libring-test.a" "-l" "stdc++" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libtest-a4729905.rlib" "/Users/mbp/dev/forks/ring/target/debug/deps/liblazy_static-7f1b96a3a3eb529d.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libgetopts-a4729905.rlib" "/Users/mbp/dev/forks/ring/target/debug/deps/libuntrusted-6d5be7309ea48309.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libterm-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libstd-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libpanic_unwind-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libunwind-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcollections-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liballoc-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liballoc_jemalloc-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/liblibc-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librand-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/librustc_unicode-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcore-a4729905.rlib" "/Users/mbp/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/x86_64-apple-darwin/lib/libcompiler_builtins-a4729905.rlib" "-l" "System" "-l" "pthread" "-l" "c" "-l" "m"
  = note: Undefined symbols for architecture x86_64:
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::basic_string(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      std::__1::unique_ptr<std::__1::__tree_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*>, std::__1::__tree_node_destructor<std::__1::allocator<std::__1::__tree_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*> > > > std::__1::__tree<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::__construct_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&&&) in libring-test.a(file_test.o)
      std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >::__construct_node_with_key(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libring-test.a(file_test.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::operator=(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      FileTest::ReadNext() in libring-test.a(file_test.o)
      FileTest::GetAttribute(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libring-test.a(file_test.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__init(char const*, unsigned long)", referenced from:
      GetBIGNUM(FileTest*, char const*) in libring-test.a(bn_test.o)
      StripSpace(char const*, unsigned long) in libring-test.a(file_test.o)
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_string()", referenced from:
      GetBIGNUM(FileTest*, char const*) in libring-test.a(bn_test.o)
      FileTest::FileTest(char const*) in libring-test.a(file_test.o)
      FileTest::~FileTest() in libring-test.a(file_test.o)
      FileTest::ReadNext() in libring-test.a(file_test.o)
      std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >::operator[](std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in libring-test.a(file_test.o)
      std::__1::__tree<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > >::destroy(std::__1::__tree_node<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, void*>*) in libring-test.a(file_test.o)
      std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::~pair() in libring-test.a(file_test.o)
      ...
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::compare(char const*) const", referenced from:
      RunTest(FileTest*, void*) in libring-test.a(bn_test.o)
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)


error: aborting due to previous error

error: Could not compile `ring`.

Caused by:
  process didn't exit successfully: `rustc src/lib.rs --crate-name ring -g --test --cfg feature="use_heap" --cfg feature="default" --cfg feature="dev_urandom_fallback" -C metadata=2a96e7777e87260e -C extra-filename=-2a96e7777e87260e --out-dir /Users/mbp/dev/forks/ring/target/debug/deps --emit=dep-info,link -L dependency=/Users/mbp/dev/forks/ring/target/debug/deps --extern lazy_static=/Users/mbp/dev/forks/ring/target/debug/deps/liblazy_static-7f1b96a3a3eb529d.rlib --extern untrusted=/Users/mbp/dev/forks/ring/target/debug/deps/libuntrusted-6d5be7309ea48309.rlib -L native=/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/out/lib -l static=ring-core -l static=ring-test -l dylib=stdc++` (exit code: 101)
```

XCode is installed, has its command line tools, and is up to date.

`cc -version`:

```
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin16.3.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

The `cargo:` lines in `target/debug/build/ring-5ae0aa2ea1993729/output` are:

```
cargo:rustc-link-search=native=/Users/mbp/dev/forks/ring/target/debug/build/ring-5ae0aa2ea1993729/out/lib
cargo:rustc-link-lib=static=ring-core
cargo:rustc-link-lib=static=ring-test
cargo:rustc-flags=-l dylib=stdc++
```

Let me know if there are other parts of the output that would be helpful.

**Comment by briansmith** [comment]
Did it every work for you, using any version of *ring* on any version of macOS using any version of xcode?

Try removing the C++-based tests from the build and see if that works around the problem:

```diff
 #[cfg(test)]
 mod tests {
     #[cfg(feature = "use_heap")]
-     bssl_test_rng!(test_bn, bssl_bn_test_main);
 
-     bssl_test!(test_constant_time, bssl_constant_time_test_main);
 }
```

**Comment by marshallpierce** [comment]
Removing those two tests yields a passing test run.

It has never worked for me; my earlier *ring* work was done on a linux box so I only just discovered this when I was traveling. I have two 10.12 laptops available and they both exhibit the same behavior. (One of them is having GPU troubles today, so let's just say it didn't work there yesterday but I can't really make further attempts.) XCode is at version 8.2.

8b7bb0cf6b649870bdf3150a3122e6ebe2962d24 (0.4.2) exhibits the same behavior as current master (53fd7f97). 
5778edbb5a672266a2307ebf7377880568f36ddf (0.4.1), the previous commit, does not compile, but because of rust issues (as evidenced by 0.4.2's commit message).

**Comment by briansmith** [comment]
My current hypothesis is that there is some breaking change between xcode 7.3 (used on Travis CI) and xcode 8.2. I've submitted a test build to see: https://travis-ci.org/briansmith/ring/builds/185929700

**Comment by briansmith** [comment]
> My current hypothesis is that there is some breaking change between xcode 7.3 (used on Travis CI) and xcode 8.2. I've submitted a test build to see: https://travis-ci.org/briansmith/ring/builds/185929700

Confirmed: https://travis-ci.org/briansmith/ring/jobs/185929701#L225

We may need to change from this:
```cargo:rustc-flags=-l dylib=stdc++```
```cargo:rustc-flags=-l dylib=c++```

**Comment by briansmith** [comment]
> We may need to change from this:
> cargo:rustc-flags=-l dylib=stdc++
> cargo:rustc-flags=-l dylib=c++

I submitted a change to test this: https://travis-ci.org/briansmith/ring/builds/185938500. When targetting Mac OS X I suppose we should always link to libc++. However, how are we supposed to know whether to link to libc++ or libstdc++ on Linux?

**Comment by briansmith** [comment]
Fixed via #386 in 0.6.1, which I just published. Thanks for reporting this bug. I was completely oblivious to the problem.
